File: sdk/python/src/basanos_sdk/transport.py
"""Small JSON-RPC transport for the DCG SDK."""
from __future__ import annotations

import base64
import json
import os
import time
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Protocol, Self

import httpx
from solders.hash import Hash
from solders.instruction import Instruction
from solders.keypair import Keypair
from solders.message import Message
from solders.pubkey import Pubkey
from solders.transaction import Transaction

from . import consensus


class RpcError(RuntimeError):
    """A JSON-RPC error response."""


class RpcBodyTooLarge(ValueError):
    """The request cannot fit the endpoint's 50 KiB JSON body limit."""


class Signer(Protocol):
    @property
    def pubkey(self) -> Pubkey:
        ...

    def sign_transaction(self, transaction: Transaction) -> None:
        ...


@dataclass(frozen=True)
class Account:
    pubkey: Pubkey
    owner: Pubkey | None
    lamports: int
    data: bytes
    executable: bool
    rent_epoch: int | None
    space: int | None


@dataclass(frozen=True)
class Confirmation:
    signature: str
    commitment: str
    slot: int | None
    err: Any
    status: str


@dataclass(frozen=True)
class SendReceipt:
    signature: str
    rpc_response: dict[str, Any]


def load_keypair(path: str | os.PathLike[str]) -> Keypair:
    """Load a JSON byte-array or base58 key file without exposing its contents."""
    location = Path(path)
    mode = location.stat().st_mode
    if mode & 0o077:
        raise ValueError("key file must not be group or world accessible")
    raw = location.read_text().strip()
    try:
        value = json.loads(raw)
        if isinstance(value, list):
            return Keypair.from_bytes(bytes(value))
        if isinstance(value, str):
            return Keypair.from_base58_string(value)
    except (json.JSONDecodeError, TypeError, ValueError):
        pass
    try:
        return Keypair.from_base58_string(raw)
    except ValueError as error:
        raise ValueError("invalid key file") from error


def signer(value: Keypair | str | os.PathLike[str]) -> Keypair:
    return value if isinstance(value, Keypair) else load_keypair(value)


def signer_pubkey(value: Keypair | str | os.PathLike[str]) -> Pubkey:
    return signer(value).pubkey()


class RpcClient:
    """A bounded JSON-RPC client for one Solana-compatible endpoint."""

    def __init__(self, endpoint: str, *, client: Any | None = None, timeout: float = 30.0,
                 commitment: str = "confirmed", max_body_bytes: int = consensus.JSON_RPC_BODY_LIMIT) -> None:
        if not endpoint or not isinstance(endpoint, str):
            raise ValueError("endpoint must be a URL string")
        if max_body_bytes < 1 or max_body_bytes > consensus.JSON_RPC_BODY_LIMIT:
            raise ValueError("max_body_bytes must be within the 50 KiB limit")
        self.endpoint = endpoint
        self.commitment = commitment
        self.max_body_bytes = max_body_bytes
        self._client = client or httpx.Client(timeout=timeout)
        self._owns_client = client is None
        self._request_id = 0

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> Self:
        return self

    def __exit__(self, *_args: object) -> None:
        self.close()

    def _body(self, method: str, params: Sequence[Any]) -> bytes:
        self._request_id += 1
        body = json.dumps({"jsonrpc": "2.0", "id": self._request_id, "method": method, "params": list(params)},
                          separators=(",", ":"), ensure_ascii=True).encode("utf-8")
        if len(body) > self.max_body_bytes:
            raise RpcBodyTooLarge(f"{method} request is {len(body)} bytes; limit is {self.max_body_bytes}")
        return body

    def call(self, method: str, params: Sequence[Any] = ()) -> Any:
        body = self._body(method, params)
        response = self._client.post(self.endpoint, content=body,
                                     headers={"Content-Type": "application/json"})
        if hasattr(response, "raise_for_status"):
            response.raise_for_status()
        value = response.json() if hasattr(response, "json") else json.loads(response.content)
        if not isinstance(value, dict):
            raise RpcError("JSON-RPC response is not an object")
        if value.get("error") is not None:
            error = value["error"]
            if isinstance(error, dict):
                message = str(error.get("message", "RPC error"))
                code = error.get("code")
                raise RpcError(f"{code}: {message}" if code is not None else message)
            raise RpcError(str(error))
        if "result" not in value:
            raise RpcError("JSON-RPC response has no result")
        return value["result"]

    def get_account_info(self, pubkey: Pubkey | str, *, commitment: str | None = None) -> Account | None:
        value = self.call("getAccountInfo", [str(consensus.key(pubkey)), {"encoding": "base64",
                                                                            "commitment": commitment or self.commitment}])
        return self._decode_account(str(consensus.key(pubkey)), value)

    def get_multiple_accounts(self, pubkeys: Sequence[Pubkey | str], *, commitment: str | None = None) -> list[Account | None]:
        keys = [consensus.key(value) for value in pubkeys]
        result: list[Account | None] = []
        start = 0
        batch_size = min(100, len(keys))
        while start < len(keys):
            while True:
                try:
                    values = self.call("getMultipleAccounts", [[str(value) for value in keys[start:start + batch_size]],
                                                                 {"encoding": "base64", "commitment": commitment or self.commitment}])
                    break
                except RpcBodyTooLarge:
                    if batch_size == 1:
                        raise
                    batch_size = max(1, batch_size // 2)
            if len(values) != min(batch_size, len(keys) - start):
                raise RpcError("account batch response is malformed")
            result.extend(self._decode_account(str(keys[index]), value) for index, value in enumerate(values, start))
            start += batch_size
        return result

    def read_account(self, pubkey: Pubkey | str, *, commitment: str | None = None) -> Account | None:
        return self.get_account_info(pubkey, commitment=commitment)

    def read_accounts(self, pubkeys: Sequence[Pubkey | str], *, commitment: str | None = None) -> list[Account | None]:
        return self.get_multiple_accounts(pubkeys, commitment=commitment)

    @staticmethod
    def _decode_account(address: str, value: Any) -> Account | None:
        if value is None:
            return None
        if not isinstance(value, dict):
            raise RpcError("account response is malformed")
        encoded = value.get("data")
        if isinstance(encoded, list):
            if len(encoded) != 2 or encoded[1] != "base64":
                raise RpcError("unsupported account encoding")
            data = base64.b64decode(encoded[0], validate=True)
        elif isinstance(encoded, str):
            data = base64.b64decode(encoded, validate=True)
        else:
            raise RpcError("account data is missing")
        owner = value.get("owner")
        return Account(Pubkey.from_string(address), Pubkey.from_string(owner) if owner else None,
                       int(value.get("lamports", 0)), data, bool(value.get("executable", False)),
                       int(value["rentEpoch"]) if value.get("rentEpoch") is not None else None,
                       int(value["space"]) if value.get("space") is not None else None)

    def latest_blockhash(self, *, commitment: str | None = None) -> Hash:
        value = self.call("getLatestBlockhash", [{"commitment": commitment or self.commitment}])
        if not isinstance(value, dict) or not isinstance(value.get("blockhash"), str):
            raise RpcError("latest blockhash response is malformed")
        return Hash.from_string(value["blockhash"])

    def send_raw_transaction(self, transaction: bytes | str, *, skip_preflight: bool = False,
                             max_retries: int = 0) -> SendReceipt:
        encoded = transaction if isinstance(transaction, str) else base64.b64encode(transaction).decode("ascii")
        last: Exception | None = None
        for _attempt in range(max_retries + 1):
            try:
                signature = self.call("sendTransaction", [encoded, {"encoding": "base64", "skipPreflight": skip_preflight,
                                                                      "preflightCommitment": self.commitment}])
                if not isinstance(signature, str):
                    raise RpcError("sendTransaction response is malformed")
                return SendReceipt(signature, {"signature": signature})
            except RpcError as error:
                last = error
                if not _attempt:
                    continue
        raise last or RpcError("send failed")

    def build_transaction(self, instructions: Sequence[Instruction], payer: Keypair | str | os.PathLike[str],
                          signers: Sequence[Keypair | str | os.PathLike[str]] = (), *, blockhash: Hash | None = None) -> Transaction:
        payer_key = signer(payer)
        blockhash = blockhash or self.latest_blockhash()
        message = Message.new_with_blockhash(list(instructions), payer_key.pubkey(), blockhash)
        transaction = Transaction.new_unsigned(message)
        signer_keys = [signer(value) for value in signers]
        all_signers = [payer_key, *(value for value in signer_keys if value.pubkey() != payer_key.pubkey())]
        transaction.partial_sign(all_signers, blockhash)
        return transaction

    def send_transaction(self, instructions: Sequence[Instruction], payer: Keypair | str | os.PathLike[str],
                         signers: Sequence[Keypair | str | os.PathLike[str]] = (), *, skip_preflight: bool = False,
                         max_retries: int = 0) -> SendReceipt:
        transaction = self.build_transaction(instructions, payer, signers)
        return self.send_raw_transaction(bytes(transaction), skip_preflight=skip_preflight, max_retries=max_retries)

    def signature_statuses(self, signature: str) -> list[dict[str, Any]]:
        value = self.call("getSignatureStatuses", [[signature], {"searchTransactionHistory": True}])
        if not isinstance(value, dict) or not isinstance(value.get("value"), list):
            raise RpcError("signature status response is malformed")
        return value["value"]

    def confirm(self, signature: str, *, commitment: str = "confirmed", timeout: float = 60.0,
                poll_interval: float = 0.25, clock: Callable[[], float] = time.monotonic,
                sleep: Callable[[float], None] = time.sleep) -> Confirmation:
        deadline = clock() + timeout
        order = {"processed": 0, "confirmed": 1, "finalized": 2}
        target = order[commitment]
        while True:
            statuses = self.signature_statuses(signature)
            value = statuses[0] if statuses else None
            if value is not None:
                current = value.get("confirmationStatus")
                if current is None:
                    current = "processed"
                if order.get(current, 0) >= target:
                    return Confirmation(signature, current, value.get("slot"), value.get("err"), current)
                if value.get("err") is not None:
                    return Confirmation(signature, current, value.get("slot"), value.get("err"), current)
            if clock() >= deadline:
                raise TimeoutError("transaction confirmation timed out")
            sleep(poll_interval)

    def _transaction_value(self, signature: str, *, search_history: bool) -> dict[str, Any] | None:
        value = self.call("getTransaction", [signature, {"encoding": "json", "searchTransactionHistory": search_history}])
        return value if isinstance(value, dict) else None

    def transaction_logs(self, signature: str, *, search_history: bool = True) -> list[str]:
        value = self._transaction_value(signature, search_history=search_history)
        if value is None:
            return []
        meta = ((value.get("meta") or {}).get("logMessages") or [])
        return [str(item) for item in meta]

    def logs(self, signature: str, *, search_history: bool = True) -> list[str]:
        return self.transaction_logs(signature, search_history=search_history)

    def get_logs(self, signature: str, *, search_history: bool = True) -> list[str]:
        return self.transaction_logs(signature, search_history=search_history)

    def events_from_logs(self, signature: str, *, program: Pubkey | str | None = None) -> list[dict[str, Any]]:
        value = self._transaction_value(signature, search_history=True)
        if value is None or (value.get("meta") or {}).get("err") is not None:
            return []
        meta = (value.get("meta") or {}).get("logMessages") or []
        events = []
        program_text = str(consensus.key(program)) if program is not None else None
        stack: list[str] = []
        for line in (str(item) for item in meta):
            if line.startswith("Program ") and " invoke [" in line:
                stack.append(line.split()[1])
            elif line.startswith("Program ") and (" success" in line or " failed:" in line):
                if stack:
                    stack.pop()
            if "Program data:" not in line:
                continue
            current_program = stack[-1] if stack else None
            if program_text is not None and current_program != program_text:
                continue
            _prefix, encoded = line.split("Program data:", 1)
            try:
                raw = base64.b64decode(encoded.strip(), validate=True)
                events.append(consensus.decode_event_any(raw))
            except (ValueError, TypeError):
                continue
        return events

File: sdk/typescript/src/transport.ts
import { readFileSync, statSync } from "node:fs";
import type { PathLike } from "node:fs";
import { Keypair, PublicKey, Transaction, TransactionInstruction } from "@solana/web3.js";
import bs58 from "bs58";
import * as c from "./consensus.js";

export class RpcError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RpcError";
  }
}

export class RpcBodyTooLarge extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RpcBodyTooLarge";
  }
}

export interface Account {
  pubkey: PublicKey;
  owner: PublicKey | null;
  lamports: number;
  data: Buffer;
  executable: boolean;
  rentEpoch: number | null;
  space: number | null;
}

export interface Confirmation {
  signature: string;
  commitment: string;
  slot: number | null;
  err: unknown;
  status: string;
}

export interface SendReceipt {
  signature: string;
  rpcResponse: Record<string, unknown>;
}

export interface HttpResponseLike {
  ok?: boolean;
  status?: number;
  statusText?: string;
  json?: () => Promise<unknown> | unknown;
  text?: () => Promise<string> | string;
  content?: Uint8Array | string;
  raiseForStatus?: () => void;
}

export interface HttpClientLike {
  post: (url: string, init: { body: Uint8Array; headers: Record<string, string> }) => PromiseLike<HttpResponseLike> | HttpResponseLike;
  close?: () => void;
}

export interface RpcClientOptions {
  client?: HttpClientLike;
  fetch?: typeof globalThis.fetch;
  timeout?: number;
  commitment?: string;
  maxBodyBytes?: number;
}

export function loadKeypair(path: string | PathLike): Keypair {
  const location = String(path);
  return loadKeypairSync(location);
}

function loadKeypairSync(location: string): Keypair {
  const mode = statSync(location).mode;
  if ((mode & 0o077) !== 0) throw new Error("key file must not be group or world accessible");
  const raw = readFileSync(location, "utf8").trim();
  try {
    const value: unknown = JSON.parse(raw);
    if (Array.isArray(value)) return Keypair.fromSecretKey(Uint8Array.from(value));
    if (typeof value === "string") return Keypair.fromSecretKey(bs58.decode(value));
  } catch {
    void 0;
  }
  try {
    return Keypair.fromSecretKey(bs58.decode(raw));
  } catch {
    throw new Error("invalid key file");
  }
}

export function signer(value: Keypair | string | PathLike): Keypair {
  return value instanceof Keypair ? value : loadKeypair(value);
}

export function signerPubkey(value: Keypair | string | PathLike): PublicKey {
  return signer(value).publicKey;
}

export const load_keypair = loadKeypair;
export const signer_pubkey = signerPubkey;

type FetchClient = {
  post: (url: string, init: { body: Uint8Array; headers: Record<string, string> }) => PromiseLike<HttpResponseLike> | HttpResponseLike;
  close?: () => void;
};

function decodeBase64(value: string): Buffer {
  if (value.length % 4 !== 0 || !/^[A-Za-z0-9+/]*={0,2}$/.test(value)) throw new RpcError("account data is not valid base64");
  return Buffer.from(value, "base64");
}

export class RpcClient {
  readonly endpoint: string;
  readonly commitment: string;
  readonly maxBodyBytes: number;
  private readonly client: FetchClient;
  private readonly ownsClient: boolean;
  private requestId = 0;

  constructor(endpoint: string, options: RpcClientOptions = {}) {
    if (typeof endpoint !== "string" || endpoint.length === 0) throw new Error("endpoint must be a URL string");
    const maxBodyBytes = options.maxBodyBytes ?? c.JSON_RPC_BODY_LIMIT;
    if (!Number.isSafeInteger(maxBodyBytes) || maxBodyBytes < 1 || maxBodyBytes > c.JSON_RPC_BODY_LIMIT) throw new Error("max_body_bytes must be within the 50 KiB limit");
    this.endpoint = endpoint;
    this.commitment = options.commitment ?? "confirmed";
    this.maxBodyBytes = maxBodyBytes;
    if (options.client !== undefined) {
      this.client = options.client;
      this.ownsClient = false;
    } else {
      const fetchFunction = options.fetch ?? globalThis.fetch;
      if (typeof fetchFunction !== "function") throw new Error("a fetch implementation is required");
      this.client = {
        post: async (url, init) => {
          const controller = new AbortController();
          const timer = options.timeout === undefined ? undefined : setTimeout(() => controller.abort(), options.timeout * 1000);
          try {
            return await fetchFunction(url, { method: "POST", body: init.body as unknown as BodyInit, headers: init.headers, signal: controller.signal });
          } finally {
            if (timer !== undefined) clearTimeout(timer);
          }
        },
      };
      this.ownsClient = true;
    }
  }

  close(): void {
    if (this.ownsClient) this.client.close?.();
  }

  private body(method: string, params: readonly unknown[]): Buffer {
    this.requestId += 1;
    const encoded = Buffer.from(JSON.stringify({ jsonrpc: "2.0", id: this.requestId, method, params }), "utf8");
    if (encoded.length > this.maxBodyBytes) throw new RpcBodyTooLarge(`${method} request is ${encoded.length} bytes; limit is ${this.maxBodyBytes}`);
    return encoded;
  }

  async call(method: string, params: readonly unknown[] = []): Promise<unknown> {
    const body = this.body(method, params);
    const response = await this.client.post(this.endpoint, { body, headers: { "Content-Type": "application/json" } });
    if (response.ok === false) throw new RpcError(`HTTP ${response.status ?? ""} ${response.statusText ?? ""}`.trim());
    if (response.raiseForStatus !== undefined) response.raiseForStatus();
    let value: unknown;
    if (response.json !== undefined) value = await response.json();
    else if (response.text !== undefined) value = JSON.parse(await response.text());
    else if (response.content !== undefined) value = JSON.parse(Buffer.from(response.content).toString("utf8"));
    else throw new RpcError("JSON-RPC response is not readable");
    if (typeof value !== "object" || value === null || Array.isArray(value)) throw new RpcError("JSON-RPC response is not an object");
    const object = value as Record<string, unknown>;
    if (object.error !== undefined && object.error !== null) {
      const error = object.error;
      if (typeof error === "object" && error !== null) {
        const message = String((error as Record<string, unknown>).message ?? "RPC error");
        const code = (error as Record<string, unknown>).code;
        throw new RpcError(code === undefined ? message : `${String(code)}: ${message}`);
      }
      throw new RpcError(String(error));
    }
    if (!("result" in object)) throw new RpcError("JSON-RPC response has no result");
    return object.result;
  }

  async getAccountInfo(pubkey: PublicKey | string | Uint8Array, commitment?: string): Promise<Account | null> {
    const address = c.key(pubkey);
    const value = await this.call("getAccountInfo", [address.toBase58(), { encoding: "base64", commitment: commitment ?? this.commitment }]);
    return RpcClient.decodeAccount(address, value);
  }

  async getMultipleAccounts(pubkeys: readonly (PublicKey | string | Uint8Array)[], commitment?: string): Promise<Array<Account | null>> {
    const keys = pubkeys.map((value) => c.key(value));
    const result: Array<Account | null> = [];
    let start = 0;
    let batchSize = Math.min(100, keys.length);
    while (start < keys.length) {
      let values: unknown;
      while (true) {
        try {
          values = await this.call("getMultipleAccounts", [keys.slice(start, start + batchSize).map((value) => value.toBase58()), { encoding: "base64", commitment: commitment ?? this.commitment }]);
          break;
        } catch (error) {
          if (!(error instanceof RpcBodyTooLarge) || batchSize === 1) throw error;
          batchSize = Math.max(1, Math.floor(batchSize / 2));
        }
      }
      const expected = Math.min(batchSize, keys.length - start);
      if (!Array.isArray(values) || values.length !== expected) throw new RpcError("account batch response is malformed");
      values.forEach((value, index) => result.push(RpcClient.decodeAccount(keys[start + index], value)));
      start += batchSize;
    }
    return result;
  }

  async readAccount(pubkey: PublicKey | string | Uint8Array, commitment?: string): Promise<Account | null> { return this.getAccountInfo(pubkey, commitment); }
  async readAccounts(pubkeys: readonly (PublicKey | string | Uint8Array)[], commitment?: string): Promise<Array<Account | null>> { return this.getMultipleAccounts(pubkeys, commitment); }
  async get_account_info(pubkey: PublicKey | string | Uint8Array, commitment?: string): Promise<Account | null> { return this.getAccountInfo(pubkey, commitment); }
  async get_multiple_accounts(pubkeys: readonly (PublicKey | string | Uint8Array)[], commitment?: string): Promise<Array<Account | null>> { return this.getMultipleAccounts(pubkeys, commitment); }
  async read_account(pubkey: PublicKey | string | Uint8Array, commitment?: string): Promise<Account | null> { return this.readAccount(pubkey, commitment); }
  async read_accounts(pubkeys: readonly (PublicKey | string | Uint8Array)[], commitment?: string): Promise<Array<Account | null>> { return this.readAccounts(pubkeys, commitment); }

  static decodeAccount(address: PublicKey, value: unknown): Account | null {
    if (value === null || value === undefined) return null;
    if (typeof value !== "object" || Array.isArray(value)) throw new RpcError("account response is malformed");
    const object = value as Record<string, unknown>;
    const encoded = object.data;
    let data: Buffer;
    if (Array.isArray(encoded) && encoded.length === 2 && encoded[1] === "base64") data = decodeBase64(String(encoded[0]));
    else if (typeof encoded === "string") data = decodeBase64(encoded);
    else throw new RpcError("account data is missing");
    const owner = object.owner === undefined || object.owner === null ? null : new PublicKey(String(object.owner));
    const lamportsNumber = Number(object.lamports ?? 0);
    if (!Number.isSafeInteger(lamportsNumber) || lamportsNumber < 0) throw new RpcError("account lamports are malformed");
    const rentEpoch = object.rentEpoch === undefined || object.rentEpoch === null ? null : Number(object.rentEpoch);
    const space = object.space === undefined || object.space === null ? null : Number(object.space);
    if ((rentEpoch !== null && !Number.isSafeInteger(rentEpoch)) || (space !== null && !Number.isSafeInteger(space))) throw new RpcError("account metadata is malformed");
    return { pubkey: address, owner, lamports: lamportsNumber, data, executable: Boolean(object.executable), rentEpoch, space };
  }

  async latestBlockhash(commitment?: string): Promise<string> {
    const value = await this.call("getLatestBlockhash", [{ commitment: commitment ?? this.commitment }]);
    if (typeof value !== "object" || value === null || typeof (value as Record<string, unknown>).blockhash !== "string") throw new RpcError("latest blockhash response is malformed");
    return (value as Record<string, string>).blockhash;
  }

  async sendRawTransaction(transaction: Uint8Array | string, options: { skipPreflight?: boolean; maxRetries?: number } = {}): Promise<SendReceipt> {
    const encoded = typeof transaction === "string" ? transaction : Buffer.from(transaction).toString("base64");
    const maxRetries = options.maxRetries ?? 0;
    let last: unknown;
    for (let attempt = 0; attempt <= maxRetries; attempt += 1) {
      try {
        const signature = await this.call("sendTransaction", [encoded, { encoding: "base64", skipPreflight: options.skipPreflight ?? false, preflightCommitment: this.commitment }]);
        if (typeof signature !== "string") throw new RpcError("sendTransaction response is malformed");
        return { signature, rpcResponse: { signature } };
      } catch (error) {
        last = error;
      }
    }
    throw last instanceof Error ? last : new RpcError("send failed");
  }

  async buildTransaction(instructions: readonly TransactionInstruction[], payer: Keypair | string | PathLike, signers: readonly (Keypair | string | PathLike)[] = [], blockhash?: string): Promise<Transaction> {
    const payerKey = signer(payer);
    const recentBlockhash = blockhash ?? await this.latestBlockhash();
    const transaction = new Transaction({ feePayer: payerKey.publicKey, recentBlockhash });
    transaction.add(...instructions);
    const unique = new Map<string, Keypair>();
    unique.set(payerKey.publicKey.toBase58(), payerKey);
    for (const value of signers) {
      const key = signer(value);
      if (!unique.has(key.publicKey.toBase58())) unique.set(key.publicKey.toBase58(), key);
    }
    transaction.partialSign(...unique.values());
    return transaction;
  }

  async sendTransaction(instructions: readonly TransactionInstruction[], payer: Keypair | string | PathLike, signers: readonly (Keypair | string | PathLike)[] = [], options: { skipPreflight?: boolean; maxRetries?: number } = {}): Promise<SendReceipt> {
    const transaction = await this.buildTransaction(instructions, payer, signers);
    return this.sendRawTransaction(transaction.serialize(), options);
  }

  async send_raw_transaction(transaction: Uint8Array | string, options: { skipPreflight?: boolean; maxRetries?: number } = {}): Promise<SendReceipt> { return this.sendRawTransaction(transaction, options); }
  async build_transaction(instructions: readonly TransactionInstruction[], payer: Keypair | string | PathLike, signers: readonly (Keypair | string | PathLike)[] = [], blockhash?: string): Promise<Transaction> { return this.buildTransaction(instructions, payer, signers, blockhash); }
  async send_transaction(instructions: readonly TransactionInstruction[], payer: Keypair | string | PathLike, signers: readonly (Keypair | string | PathLike)[] = [], options: { skipPreflight?: boolean; maxRetries?: number } = {}): Promise<SendReceipt> { return this.sendTransaction(instructions, payer, signers, options); }

  async signatureStatuses(signature: string): Promise<Array<Record<string, unknown>>> {
    const value = await this.call("getSignatureStatuses", [[signature], { searchTransactionHistory: true }]);
    if (typeof value !== "object" || value === null || !Array.isArray((value as Record<string, unknown>).value)) throw new RpcError("signature status response is malformed");
    return (value as { value: Array<Record<string, unknown>> }).value;
  }

  async signature_statuses(signature: string): Promise<Array<Record<string, unknown>>> { return this.signatureStatuses(signature); }

  async confirm(signature: string, options: { commitment?: string; timeout?: number; pollInterval?: number; clock?: () => number; sleep?: (seconds: number) => void } = {}): Promise<Confirmation> {
    const deadline = (options.clock ?? Date.now)() + (options.timeout ?? 60) * 1000;
    const order: Record<string, number> = { processed: 0, confirmed: 1, finalized: 2 };
    const target = order[options.commitment ?? "confirmed"] ?? 1;
    const clock = options.clock ?? Date.now;
    const sleep = options.sleep ?? ((seconds: number) => new Promise<void>((resolve) => setTimeout(resolve, seconds * 1000)));
    while (true) {
      const statuses = await this.signatureStatuses(signature);
      const value = statuses[0];
      if (value !== undefined) {
        const current = typeof value.confirmationStatus === "string" ? value.confirmationStatus : "processed";
        if ((order[current] ?? 0) >= target) return { signature, commitment: current, slot: value.slot === undefined ? null : Number(value.slot), err: value.err ?? null, status: current };
        if (value.err !== undefined && value.err !== null) return { signature, commitment: current, slot: value.slot === undefined ? null : Number(value.slot), err: value.err, status: current };
      }
      if (clock() >= deadline) throw new Error("transaction confirmation timed out");
      await sleep(options.pollInterval ?? 0.25);
    }
  }

  private async transactionValue(signature: string, searchHistory: boolean): Promise<Record<string, unknown> | null> {
    const value = await this.call("getTransaction", [signature, { encoding: "json", searchTransactionHistory: searchHistory }]);
    return typeof value === "object" && value !== null && !Array.isArray(value) ? value as Record<string, unknown> : null;
  }

  async transactionLogs(signature: string, searchHistory = true): Promise<string[]> {
    const value = await this.transactionValue(signature, searchHistory);
    if (value === null) return [];
    const meta = typeof value.meta === "object" && value.meta !== null ? value.meta as Record<string, unknown> : undefined;
    const logs = meta?.logMessages;
    return Array.isArray(logs) ? logs.map((item) => String(item)) : [];
  }

  async logs(signature: string, searchHistory = true): Promise<string[]> { return this.transactionLogs(signature, searchHistory); }
  async getLogs(signature: string, searchHistory = true): Promise<string[]> { return this.transactionLogs(signature, searchHistory); }
  async transaction_logs(signature: string, searchHistory = true): Promise<string[]> { return this.transactionLogs(signature, searchHistory); }
  async get_logs(signature: string, searchHistory = true): Promise<string[]> { return this.transactionLogs(signature, searchHistory); }

  async eventsFromLogs(signature: string, program?: PublicKey | string | Uint8Array): Promise<Array<Record<string, c.Integer | Buffer | string>>> {
    const value = await this.transactionValue(signature, true);
    if (value === null) return [];
    const meta = typeof value.meta === "object" && value.meta !== null ? value.meta as Record<string, unknown> : undefined;
    if (meta?.err !== undefined && meta.err !== null) return [];
    const logs = Array.isArray(meta?.logMessages) ? meta.logMessages.map((item) => String(item)) : [];
    const programText = program === undefined ? undefined : c.key(program).toBase58();
    let currentProgram: string | undefined;
    const events: Array<Record<string, c.Integer | Buffer | string>> = [];
    for (const line of logs) {
      if (line.startsWith("Program ") && line.includes(" invoke [")) currentProgram = line.split(/\s+/)[1];
      if (!line.includes("Program data:")) continue;
      if (programText !== undefined && currentProgram !== programText) continue;
      const encoded = line.slice(line.indexOf("Program data:") + "Program data:".length).trim();
      try { events.push(c.decodeEvent(Buffer.from(encoded, "base64"))); } catch { void 0; }
    }
    return events;
  }
  async events_from_logs(signature: string, program?: PublicKey | string | Uint8Array): Promise<Array<Record<string, c.Integer | Buffer | string>>> { return this.eventsFromLogs(signature, program); }
}

export const RpcBodyTooLargeError = RpcBodyTooLarge;

export function eventsFromLogs(rpc: RpcClient, signature: string, program?: PublicKey | string | Uint8Array): Promise<Array<Record<string, c.Integer | Buffer | string>>> {
  return rpc.eventsFromLogs(signature, program);
}

export const events_from_logs = eventsFromLogs;
